graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -238,13 +315,330 @@ def _process_generic_type_definition(generic_type, schema_id, existing_scalars,
+        type_name_to_schema_id: Dict[str, str], mapping type name to the id of the schema that the
+                                type is from
+        scalars: Set[str], names of all scalars in the merged_schema so far
+        cross_schema_edge: CrossSchemaEdgeDescriptor namedtuple, the edge that we check the
+                           validity of
+
+    Raises:
+        - InvalidCrossSchemaEdgeError if the cross schema edge lies within one schema, refers
+          to nonexistent schemas, types, or fields, refers to Union types, stitches together
+          fields that are not of a scalar type, or stitches together fields that are of
+          different scalar types
+    """
+    outbound_side = cross_schema_edge.outbound_side
+    inbound_side = cross_schema_edge.inbound_side
+
+    _check_field_reference_is_valid(type_name_to_definition, type_name_to_schema_id, outbound_side)
+    _check_field_reference_is_valid(type_name_to_definition, type_name_to_schema_id, inbound_side)
+
+    if outbound_side.schema_id == inbound_side.schema_id:  # not cross schema
+        raise InvalidCrossSchemaEdgeError(
+            u'Edge "{}" does not cross schemas.'.format(cross_schema_edge)
+        )
+
+    _check_field_types_are_matching_scalars(type_name_to_definition, scalars, cross_schema_edge)
+
+
+def _check_field_reference_is_valid(type_name_to_definition, type_name_to_schema_id,
+                                    field_reference):
+    """Check that the field reference refers to a valid field.
+
+    In particular, check that the field reference is on a type that exists in the correct
+    schema, and that the type contains the field of the expected name.
+
+    Args:
+        type_name_to_definition: Dict[str, (Interface/Object/Union)TypeDefinition], mapping
+                                 name of types to their definitions
+        type_name_to_schema_id: Dict[str, str], mapping type name to the id of the schema that the
+                                type is from
+        field_reference: FieldReference namedtuple, what we check the validity of
+
+    Raises:
+        - InvalidCrossSchemaEdgeError if the cross schema edge refers to nonexistent schemas,
+          types, or fields, or refers to Union types
+    """
+    schema_id = field_reference.schema_id
+    type_name = field_reference.type_name
+    field_name = field_reference.field_name
+
+    # Error if the type is nonexistent (includes if type is an enum or scalar)
+    # TODO: Consider improving error message if type is a union or enum or scalar
+    if type_name not in type_name_to_definition:
+        raise InvalidCrossSchemaEdgeError(
+            u'Type "{}" specified in the field reference "{}" is not found '
+            u'in the merged schema.'.format(type_name, field_reference)
+        )
+
+    # Error if the type is in a wrong or nonexistent schema
+    if type_name_to_schema_id[type_name] != schema_id:
```

Comment:
check if type_name is in type_name_to_schema_id at all?

Reply:
type_name_to_definition.keys() is a subset of type_name_to_schema_id.keys(), so the previous if statement actually ensures type_name is in type_name_to_schema_id. 

The naming is definitely not good here. type_name_to_definition only contains objects and interfaces, while type_name_to_schema_id also contains unions and enums in addition. Not sure how to change variable names to succinctly describe these differences.

Reply:
Updating the docstring to capture this subtlety would be the next best thing. Naming is hard :)